src/data_source/fetching.rs

```diff
@@ -634,10 +814,51 @@ where
         // Wait for the object to arrive.
         Fetch::Pending(fut.boxed())
     }
+
+    /// Proactively search for and retrieve missing objects.
+    ///
+    /// This function will proactively identify and retrieve blocks and leaves which are missing
+    /// from storage. It will run until cancelled, thus, it is meant to be spawned as a background
+    /// task rather than called synchronously.
+    async fn proactive_scan(self: Arc<Self>, minor_interval: Duration, major_interval: usize) {
+        let mut prev_height = 0;
+
+        for i in 0.. {
+            // Get the block height; we will look for any missing blocks up to `block_height`.
+            let block_height = { self.storage.read().await.height } as usize;
+
+            // In a major scan, we fetch all blocks between 0 and `block_height`. In minor scans
+            // (much more frequent) we fetch blocks that are missing since the last scan.
+            let start = if i % major_interval == 0 {
+                0
+            } else {
+                prev_height
+            };
+            tracing::info!("starting proactive scan {i} of blocks from {start}-{block_height}");
+            prev_height = block_height;
+
+            // Iterate over all blocks that we should have. Merely iterating over the `Fetch`es
+            // without awaiting them is enough to trigger active fetches of missing blocks, since
+            // we always trigger an active fetch when fetching by block number. Moreover, fetching
+            // the block is enough to trigger an active fetch of the corresponding leaf if it too is
+            // missing, so this one loop takes care of all resources that might be missing.
+            //
+            // The chunking behavior of `get_range` automatically ensures that, no matter how big
+            // the range is, we will release the read lock on storage every `chunk_size` items, so
```

Comment:
It may be a good idea to have the proactive fetching range chunk be a different parameter, just in case we want to tune the how greedy the background task is on the lock.

Reply:
Good idea, will do